episodes/instance-segmentation-and-measurements.md

```diff
@@ -0,0 +1,596 @@
+
+![](fig/binary_mask_erosion_10.png){
+alt="A screenshot of a semantic segmentation mask eroded with a ball of
+radius 10."}
+Erosion with a radius of 10 separates all nuclei.
+
+
+:::::::::::::::::::::::::
+:::::::::::::::::::::::::
+
+Now we have separate nuclei, lets try creating instance labels
+again.
+
+```python
+# Remove the incorrect instance segmentation from the viewer
+viewer.layers.remove('instance_seg')
+
+eroded_semantic_seg = viewer.layers['eroded ball 10'].data
+# Create a new instace segmentation using the eroded mask
+instance_seg = label(eroded_semantic_seg)
+
+viewer.add_labels(instance_seg)
+
+print(f"Number of nuclei after erosion  = {instance_seg.max()}")
+```
+
+```output
+Number of nuclei after erosion  = 19
+```
+
+![](fig/instance_segmentation_eroded.png){
+alt="Instance segmentation on the eroded segmentation mask"}
+Looking at the image above, there are no longer any incorrectly joined
+nuclei. The absolute number of nuclei found hasn't changed much as the
+erosion process has removed some partial nuclei around the edges of the
+image.
+
+Performing any size or shape analysis on these nuclei will be flawed, as
+they are heavily eroded. We can largely undo much of the erosion by using
+the scikit-images [expand labels](
+https://scikit-image.org/docs/stable/api/skimage.segmentation.html#skimage.segmentation.expand_labels)
+function.
+
+```python
+from skimage.segmentation import expand_labels
+
+# Remove the eroded instance segmentation from the viewer
+viewer.layers.remove('instance_seg')
+
+# Expand the labels using the same radius we used when eroding them (10)
+instance_seg = expand_labels(instance_seg, 10)
+
+# Put the new instance segmentation back in the viewer
+viewer.add_labels(instance_seg)
+```
+![](fig/instance_segmentation_expanded.png){
+alt="Expanded instance segmentation on the eroded segmentation mask"}
+There are now 19 apparently correctly labelled nuclei that appear to be
+the same shape as in the original mask image.
+
+:::::::::::::::::::::::::challenge
+
+### Is the erosion completely reversible?
+In order to create a correct instance segmentation we have performed a
+mask erosion followed by a label expansion. This is a common image
+operation often used to remove back ground noise, known as as `opening`,
+or an erosion followed by a dilation. In addition to helping us separate
+instances it will have the effect of removing objects smaller than the
+erosion mask, in this case a sphere with radius 10 pixels.
+If we compare the eroded and expanded image with the original mask,
+what will we see?
+
+:::::::::::::::::::::::::solution
+
+![](fig/instance_segmentation_vs_semantic_segmentation.png){
+alt="A comparison between the expanded instance segmentation and the
+original semantic segmentation showing some mismatch between the borders."}
+Looking at the above image we can see some small mismatches around the
+edges of most of the nuclei. Will the effect of this on the accuracy of
```

Comment:
But probably mentioning how much is lost on the cell at the very bottom as well.

Reply:
```suggestion
edges of most of the nuclei. It should be remembered when looking at this image that it is a single slice though a 3D image, so in some cases where the differences look large (for example the nucleus at the bottom right) they may still be only one pixel deep. Will the effect of this on the accuracy of
```